c2rust-refactor/src/scripting/to_lua_ast_node.rs

```diff
@@ -0,0 +1,362 @@
+    }
+
+    pub fn into_inner(self) -> T {
+        Arc::try_unwrap(self.0)
+            .unwrap_or_else(|_| panic!("LuaAstNode is duplicated"))
+            .into_inner()
+    }
+
+    pub fn borrow(&self) -> Ref<T> {
+        self.0.borrow()
+    }
+
+    pub fn map<F>(&self, f: F)
+        where F: Fn(&mut T)
+    {
+        f(self.0.borrow_mut().deref_mut());
+    }
+}
+
+impl<T> LuaAstNode<T>
+    where T: WalkAst
+{
+    pub fn walk<V: MutVisitor>(&self, visitor: &mut V) {
+        self.0.borrow_mut().walk(visitor);
+    }
+}
+
+unsafe impl Send for LuaAstNode<P<Item>> {}
+#[allow(unused_doc_comments)]
+impl UserData for LuaAstNode<P<Item>> {
+    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
+        methods.add_method("get_kind", |_lua_ctx, this, ()| {
+            Ok(this.0.borrow().node.ast_name())
+        });
+
+        methods.add_method("get_id", |lua_ctx, this, ()| {
+            Ok(this.0.borrow().id.to_lua(lua_ctx))
+        });
+
+        methods.add_method("get_ident", |_lua_ctx, this, ()| {
+            Ok(this.0.borrow().ident.to_string())
+        });
+
+        methods.add_method("set_ident", |_lua_ctx, this, ident: String| {
+            this.0.borrow_mut().ident = Ident::from_str(&ident);
+            Ok(())
+        });
+
+        /// Visit statements
+        // @function visit_stmts
+        // @tparam function(LuaAstNode) callback Function to call when visiting each statement
+        methods.add_method("visit_stmts", |lua_ctx, this, callback: Function| {
+            visit_nodes(&**this.borrow(), |node: &Stmt| {
+                callback.call::<_, ()>(node.clone().to_lua(lua_ctx))
+                .unwrap_or_else(|e| panic!("Lua callback failed in visit_stmts: {}", DisplayLuaError(e)));
+            });
+            Ok(())
+        });
+
+        methods.add_method("visit_items", |lua_ctx, this, callback: Function| {
+            visit_nodes(&**this.borrow(), |node: &Item| {
+                callback.call::<_, ()>(P(node.clone()).to_lua(lua_ctx))
+                .unwrap_or_else(|e| panic!("Lua callback failed in visit_items: {}", DisplayLuaError(e)));
+            });
+            Ok(())
+        });
+
+        methods.add_method("visit_foreign_items", |lua_ctx, this, callback: Function| {
+            visit_nodes(&**this.borrow(), |node: &ForeignItem| {
+                callback.call::<_, ()>(P(node.clone()).to_lua(lua_ctx))
+                .unwrap_or_else(|e| panic!("Lua callback failed in visit_foreign_items: {}", DisplayLuaError(e)));
+            });
+            Ok(())
+        });
+
+        methods.add_method("get_node", |lua_ctx, this, ()| {
+            match this.0.borrow().node.clone() {
+                ItemKind::Use(e) => Ok(e.to_lua(lua_ctx)),
+                node => Err(Error::external(format!("Item node {:?} not implemented yet", node))),
+            }
+        });
+    }
+}
+
+unsafe impl Send for LuaAstNode<P<ForeignItem>> {}
```

Comment:
Is this a good idea? I take it `ForeignItem` isn't send because of a span?

Reply:
Lots of things aren't `Send` because of `Span`. I gave up on thread-safety because the API is too unwieldily otherwise. `LuaAstNode` is not thread-safe and we cannot spawn new threads in Lua. I need to document this with a big warning in the Lua API docs.